bundle.go

```diff
@@ -0,0 +1,150 @@
+
+	"github.com/giantswarm/microerror"
+)
+
+// Bundle represents a single version bundle exposed by an authority. An
+// authority might exposes mutliple version bundles using the Capability
+// structure. Version bundles are aggregated into a merged structure represented
+// by the Aggregation structure. Also see the Aggregate function.
+type Bundle struct {
+	// Changelogs describe what changes are introduced by the version bundle. Each
+	// version bundle must have at least one changelog entry.
+	Changelogs []Changelog `json:"changelogs" yaml:"changelogs"`
+	// Components describe the components an authority exposes. Functionality of
+	// components listed here is guaranteed to be implemented in the according
+	// versions.
+	//
+	// NOTE that once this property is set it must never change again.
+	Components []Component `json:"components" yaml:"components"`
+	// Dependencies describe which components other authorities expose have to be
+	// available to be able to guarantee functionality this authority implements.
+	//
+	// NOTE that once this property is set it must never change again.
+	Dependencies []Dependency `json:"dependency" yaml:"dependency"`
+	// Deprecated defines a version bundle to be deprecated. Deprecated version
+	// bundles are not intended to be mainatined anymore. Further usage of a
+	// deprecated version bundle should be omitted.
+	Deprecated bool `json:"deprecated" yaml:"deprecated"`
+	// Time describes the time this version bundle got introduced.
+	//
+	// NOTE that once this property is set it must never change again.
+	Time time.Time `json:"time" yaml:"time"`
+	// Version describes the version of the version bundle. Versions of version
+	// bundles must be semver versions. Versions must not be duplicated. Versions
+	// should be incremented gradually.
+	//
+	// NOTE that once this property is set it must never change again.
+	Version string `json:"version" yaml:"version"`
+	// WIP describes if a version bundle is being developed. Usage of a version
+	// bundle still being developed should be omitted.
+	WIP bool `json:"wip" yaml:"wip"`
+}
+
+func (b Bundle) Validate() error {
+	if len(b.Changelogs) == 0 {
+		return microerror.Maskf(invalidBundleError, "changelogs must not be empty")
+	}
+	for _, c := range b.Changelogs {
+		err := c.Validate()
+		if err != nil {
+			return microerror.Maskf(invalidBundleError, err.Error())
+		}
+	}
+
+	if len(b.Components) == 0 {
+		return microerror.Maskf(invalidBundleError, "components must not be empty")
+	}
+	for _, c := range b.Components {
+		err := c.Validate()
+		if err != nil {
+			return microerror.Maskf(invalidBundleError, err.Error())
+		}
+	}
+
+	for _, d := range b.Dependencies {
+		err := d.Validate()
+		if err != nil {
+			return microerror.Maskf(invalidBundleError, err.Error())
+		}
+	}
+
+	var emptyTime time.Time
+	if b.Time == emptyTime {
+		return microerror.Maskf(invalidBundleError, "time must not be empty")
+	}
+
+	versionSplit := strings.Split(b.Version, ".")
```

Comment:
I reckon there's better ways to validate semver. Let's use a library here or so.

Reply:
So for bundle versions and component versions we can use the coreos lib probably. For the dependencies not, because there is this invented `== 1.2.x` format. This has to be done by us which is why I implemented on my own. Parts of the own implementation have just been there then so I reused it and I did not see the need to use both mixed up. Using a library for the dependency magic and adding our own stuff on top did not work out. WDYT?

Reply:
I reckoned there'd be some standardised way to test semver matching, might be an idea to look there, if that blocks using a library.

Reply:
I did not find any. What to do now in your opinion?

Reply:
Does https://godoc.org/github.com/blang/semver#Version.EQ fit?

Reply:
Will check. Thanks.

Reply:
For the record, we tried to use the lib but it did not work out quite well. The syntax is different. The lib matches differently and breaks the tests we have which makes it quite unreliable. The discussion revealed there is not much benefit further hustling into this direction. Thus we stay with the home grown solution for now. We can adapt later on if necessary.